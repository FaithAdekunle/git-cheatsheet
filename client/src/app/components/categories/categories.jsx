import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import CategoriesActions from '../../actions/categoriesActions';
import Category from './category';
import DeleteCategory from './deleteCategory';
import CreateOrEditCategory from './createOrEditCategory';
import CreateOrEditCommand from './createOrEditCommand';
import DeleteCommand from './deleteCommand';

export class Categories extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expandAll: false,
      keyword: '',
    };
    this.categoryBeingAddedOrEdited = { _id: '' };
    this.commandsBeingAddedOrEdited = { commands: [{ _id: '' }], categoryId: '' };
    this.toggleCategoriesExpansion = this.toggleCategoriesExpansion.bind(this);
    this.onKeywordChange = this.onKeywordChange.bind(this);
    this.hasCommandWithKeyword = this.hasCommandWithKeyword.bind(this);
    this.computeGrid = this.computeGrid.bind(this);
    this.filterCategoriesByKeyword = this.filterCategoriesByKeyword.bind(this);
    this.launchDeleteCategory = this.launchDeleteCategory.bind(this);
    this.launchAddOrEditCategory = this.launchAddOrEditCategory.bind(this);
    this.launchAddOrEditCommand = this.launchAddOrEditCommand.bind(this);
    this.deleteCategory = this.deleteCategory.bind(this);
    this.createOrEditCategory = this.createOrEditCategory.bind(this);
    this.createOrEditCommand = this.createOrEditCommand.bind(this);
    this.abortDeleteCategoryAction = this.abortDeleteCategoryAction.bind(this);
    this.abortAddOrEditCategoryAction = this.abortAddOrEditCategoryAction.bind(this);
    this.abortAddOrEditCommandAction = this.abortAddOrEditCommandAction.bind(this);
    this.togglePrivacyStatus = this.togglePrivacyStatus.bind(this);
    this.abortDeleteCommandAction = this.abortDeleteCommandAction.bind(this);
    this.deleteCategoryCommand = this.deleteCategoryCommand.bind(this);
    this.launchDeleteCommand = this.launchDeleteCommand.bind(this);
  }

  async componentDidMount() {
    const { fetchCategories } = this.props;
    await fetchCategories();
  }

  onKeywordChange(event) {
    const { target } = event;
    const state = { keyword: target.value };
    if (target.value.trim()) state.expandAll = true;
    this.setState(state);
  }

  abortDeleteCategoryAction() {
    this.setState({ categoryToBeDeleted: null });
  }

  abortDeleteCommandAction() {
    this.setState({ commandToBeDeleted: null });
  }

  abortAddOrEditCategoryAction(categoryBeingAddedOrEdited) {
    if (categoryBeingAddedOrEdited) {
      this.categoryBeingAddedOrEdited = categoryBeingAddedOrEdited;
    } else {
      this.categoryBeingAddedOrEdited = { _id: '' };
    }
    this.setState({ categoryToBeEdited: null });
  }

  abortAddOrEditCommandAction(commandsBeingAddedOrEdited) {
    if (commandsBeingAddedOrEdited) {
      this.commandsBeingAddedOrEdited = commandsBeingAddedOrEdited;
    } else {
      this.commandsBeingAddedOrEdited = { commands: [{ _id: '' }], categoryId: '' };
    }
    this.setState({ commandsToBeEdited: null });
  }

  createOrEditCategory(category) {
    const { createOrEditCategory, user } = this.props;
    createOrEditCategory(category, user.token).then(() => this.abortAddOrEditCategoryAction());
  }

  createOrEditCommand(commandsData) {
    const { commands, categoryId } = commandsData;
    const { editCategoryCommand, createCategoryCommands, user } = this.props;
    if (commands.length === 1 && commands[0]._id) {
      editCategoryCommand(commands[0], user.token).then(() => this.abortAddOrEditCommandAction());
    } else {
      createCategoryCommands(commands, categoryId, user.token)
        .then(() => this.abortAddOrEditCommandAction());
    }
  }

  launchDeleteCategory(category) {
    this.setState({ categoryToBeDeleted: { ...category } });
  }

  launchDeleteCommand(commandToBeDeleted) {
    this.setState({ commandToBeDeleted });
  }

  launchAddOrEditCategory(categoryToBeEdited = {
    title: '',
    privacyStatus: false,
    commands: [{ script: '', description: '', keywords: '', key: Date.now() }],
  }) {
    if (categoryToBeEdited._id === this.categoryBeingAddedOrEdited._id) {
      this.setState({ categoryToBeEdited: this.categoryBeingAddedOrEdited });
    } else {
      this.setState({ categoryToBeEdited });
    }
  }

  launchAddOrEditCommand(commandsToBeEdited) {
    if (commandsToBeEdited.commands[0]._id === this.commandsBeingAddedOrEdited.commands[0]._id) {
      this.setState({ commandsToBeEdited: this.commandsBeingAddedOrEdited });
    } else {
      this.setState({ commandsToBeEdited });
    }
  }

  deleteCategory() {
    const { deleteCategory, user } = this.props;
    const { categoryToBeDeleted } = this.state;
    deleteCategory(categoryToBeDeleted._id, user.token)
      .then(() => this.abortDeleteCategoryAction());
  }

  deleteCategoryCommand() {
    const { deleteCategoryCommand, user } = this.props;
    const { commandToBeDeleted } = this.state;
    deleteCategoryCommand(commandToBeDeleted._id, commandToBeDeleted.categoryId, user.token)
      .then(() => this.abortDeleteCommandAction());
  }

  toggleCategoriesExpansion() {
    this.setState(prevState => ({ expandAll: !prevState.expandAll }));
  }

  togglePrivacyStatus(id) {
    const { user, togglePrivacyStatus } = this.props;
    togglePrivacyStatus(id, user.token);
  }

  hasCommandWithKeyword(category) {
    const { keyword } = this.state;
    if (keyword.trim()) {
      return !!category.commands
        .find(command => !!command.keywords
          .find(word => word.toLowerCase().includes(keyword.trim().toLowerCase())));
    }
    return true;
  }

  filterCategoriesByKeyword() {
    const { categories } = this.props;
    return categories.filter(category => this.hasCommandWithKeyword(category));
  }

  computeGrid() {
    const categories = this.filterCategoriesByKeyword();
    let remainder = categories.length % 3;
    const factor = Math.floor((categories.length - remainder) / 3);
    return [...Array(3)].map(() => {
      if (remainder) {
        remainder -= 1;
        return factor + 1;
      }
      return factor;
    });
  }

  render() {
    const categories = this.filterCategoriesByKeyword();
    const {
      expandAll,
      keyword,
      categoryToBeDeleted,
      categoryToBeEdited,
      commandsToBeEdited,
      commandToBeDeleted,
    } = this.state;
    const { user } = this.props;
    const [col1, col2] = this.computeGrid();
    return (
      <React.Fragment>
        <div className="categories-component">
          {
            this.props.categories.length ? (
              <div className="command-keywords-search-container row justify-content-md-center">
                <div className="command-keyword-search col-md-4 col-lg-3">
                  <div className="input-group mb-3">
                    <div className="input-group-prepend">
                      <span className="input-group-text" id="keyword-search">
                        <FontAwesomeIcon icon="search" />
                      </span>
                    </div>
                    <input
                      type="text"
                      value={keyword}
                      onChange={this.onKeywordChange}
                      className="form-control command-keyword-search-field"
                      placeholder="enter command keyword"
                      aria-label="Username"
                      aria-describedby="keyword-search"
                    />
                  </div>
                </div>
              </div>
            ) : ''
          }
          <div className="container">
            <div className="categories-actions">
              {
                categories.length ? (
                  <button
                    className="expand-or-collapse"
                    type="button"
                    onClick={this.toggleCategoriesExpansion}
                  >
                    {expandAll ? 'Collapse All' : 'Expand All'}
                  </button>
                ) : ''
              }
              {
                user.id ? (
                  <button
                    className="launch-add-category-button"
                    type="button"
                    onClick={() => this.launchAddOrEditCategory()}
                  >
                    Add Category
                  </button>
                ) : ''
              }
            </div>
            {
              categories.length ? (
                <div className="category-components row">
                  <div className="col-md-4">
                    {
                      categories
                        .slice(0, col1).map(category => (
                          <div className="category-component" key={category._id}>
                            <Category
                              category={category}
                              expandAll={expandAll}
                              keyword={keyword}
                              user={user}
                              launchDeleteCategory={this.launchDeleteCategory}
                              launchEditCategory={this.launchAddOrEditCategory}
                              launchAddOrEditCommand={this.launchAddOrEditCommand}
                              launchDeleteCommand={this.launchDeleteCommand}
                              togglePrivacyStatus={this.togglePrivacyStatus}
                            />
                          </div>
                        ))
                    }
                  </div>
                  <div className="col-md-4">
                    {
                      categories
                        .slice(col1, col1 + col2).map(category => (
                          <div className="category-component" key={category._id}>
                            <Category
                              category={category}
                              expandAll={expandAll}
                              keyword={keyword}
                              user={user}
                              launchDeleteCategory={this.launchDeleteCategory}
                              launchEditCategory={this.launchAddOrEditCategory}
                              launchAddOrEditCommand={this.launchAddOrEditCommand}
                              launchDeleteCommand={this.launchDeleteCommand}
                              togglePrivacyStatus={this.togglePrivacyStatus}
                            />
                          </div>
                        ))
                    }
                  </div>
                  <div className="col-md-4">
                    {
                      categories
                        .slice(col1 + col2).map(category => (
                          <div className="category-component" key={category._id}>
                            <Category
                              category={category}
                              expandAll={expandAll}
                              keyword={keyword}
                              user={user}
                              launchDeleteCategory={this.launchDeleteCategory}
                              launchEditCategory={this.launchAddOrEditCategory}
                              launchAddOrEditCommand={this.launchAddOrEditCommand}
                              launchDeleteCommand={this.launchDeleteCommand}
                              togglePrivacyStatus={this.togglePrivacyStatus}
                            />
                          </div>
                        ))
                    }
                  </div>
                </div>
              ) : ''
            }
          </div>
        </div>
        {
          categoryToBeEdited ? (
            <CreateOrEditCategory
              abort={this.abortAddOrEditCategoryAction}
              category={categoryToBeEdited}
              createOrEditCategory={this.createOrEditCategory}
            />
          ) : ''
        }
        {
          commandsToBeEdited ? (
            <CreateOrEditCommand
              commands={commandsToBeEdited.commands}
              abort={this.abortAddOrEditCommandAction}
              createOrEditCommand={this.createOrEditCommand}
              categoryId={commandsToBeEdited.categoryId}
            />
          ) : ''
        }
        {
          categoryToBeDeleted ? (
            <DeleteCategory
              category={categoryToBeDeleted}
              abort={this.abortDeleteCategoryAction}
              deleteCategory={this.deleteCategory}
            />
          ) : ''
        }
        {
          commandToBeDeleted ? (
            <DeleteCommand
              abort={this.abortDeleteCommandAction}
              command={commandToBeDeleted}
              deleteCommand={this.deleteCategoryCommand}
            />
          ) : ''
        }
      </React.Fragment>
    );
  }
}

Categories.propTypes = {
  fetchCategories: PropTypes.func.isRequired,
  deleteCategory: PropTypes.func.isRequired,
  createOrEditCategory: PropTypes.func.isRequired,
  createCategoryCommands: PropTypes.func.isRequired,
  deleteCategoryCommand: PropTypes.func.isRequired,
  editCategoryCommand: PropTypes.func.isRequired,
  togglePrivacyStatus: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.any).isRequired,
  user: PropTypes.objectOf(PropTypes.any).isRequired,
};

export const mapStateToProps = state => ({
  categories: state.categories,
  user: state.user,
});

export const mapDispatchToProps = dispatch => ({
  fetchCategories: () => dispatch(CategoriesActions.fetchCategories()),
  deleteCategory: (id, token) => dispatch(CategoriesActions.deleteCategory(id, token)),
  togglePrivacyStatus: (id, token) => dispatch(CategoriesActions.toggleCategoryPrivacy(id, token)),
  createOrEditCategory: (category, token) => dispatch(CategoriesActions
    .createOrEditCategory(category, token)),
  editCategoryCommand: (command, token) => dispatch(CategoriesActions
    .editCategoryCommand(command, token)),
  createCategoryCommands: (commands, categoryId, token) => dispatch(CategoriesActions
    .createCategoryCommands(commands, categoryId, token)),
  deleteCategoryCommand: (commandId, categoryId, token) => dispatch(CategoriesActions
    .deleteCategoryCommand(commandId, categoryId, token)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Categories);
