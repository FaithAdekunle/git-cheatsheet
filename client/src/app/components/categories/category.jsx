import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Command from './command';
import CategoriesActions from '../../actions/categoriesActions';

export class Category extends Component {
  constructor(props) {
    super(props);
    this.state = { expand: false };
    this.toggleExpansion = this.toggleExpansion.bind(this);
    this.hasKeyWord = this.hasKeyWord.bind(this);
    this.togglePrivacyStatus = this.togglePrivacyStatus.bind(this);
  }

  componentDidMount() {
    const { expandAll } = this.props;
    this.setState({ expand: expandAll });
  }

  componentDidUpdate(prevProps) {
    const { expandAll } = this.props;
    if (expandAll !== prevProps.expandAll) this.setState({ expand: expandAll });
  }

  toggleExpansion() {
    this.setState(prevState => ({ expand: !prevState.expand }));
  }

  togglePrivacyStatus() {
    const { user, togglePrivacyStatus, category } = this.props;
    togglePrivacyStatus(category._id, user.token);
  }

  hasKeyWord(command) {
    const { keyword } = this.props;
    if (keyword.trim()) {
      return !!command.keywords
        .find(word => word.toLowerCase().includes(keyword.trim().toLowerCase()));
    }
    return true;
  }

  render() {
    const { category, user } = this.props;
    const { expand } = this.state;
    return (
      <React.Fragment>
        <div className="category-header" onClick={this.toggleExpansion}>
          <span>
            { category.title.toUpperCase() }
          </span>
        </div>
        <div className={`category-body${expand ? ' expand' : ''}`}>
          {
            user.id === category.userId ? (
              <div className="category-actions">
                <span className="privacy" onClick={this.togglePrivacyStatus}>
                  {`set to ${category.privacyStatus ? 'public' : 'private'}`}
                </span>
                <span>
                  <FontAwesomeIcon icon="pen" />
                  <FontAwesomeIcon icon="trash" />
                </span>
              </div>
            ) : ''
          }
          <ul className="command-list">
            {
              category.commands.map(command => this.hasKeyWord(command) ? (
                <li className="command-list-item" key={command._id}>
                  <Command command={command} />
                </li>
              ) : '')
            }
          </ul>
        </div>
      </React.Fragment>
    );
  }
}

Category.propTypes = {
  category: PropTypes.objectOf(PropTypes.any).isRequired,
  expandAll: PropTypes.bool.isRequired,
  keyword: PropTypes.string.isRequired,
  user: PropTypes.objectOf(PropTypes.string).isRequired,
  togglePrivacyStatus: PropTypes.func.isRequired,
};

export const mapStateToProps = () => ({});

export const mapDispatchToProps = dispatch => ({
  togglePrivacyStatus: (id, token) => dispatch(CategoriesActions.toggleCategoryPrivacy(id, token)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Category);
