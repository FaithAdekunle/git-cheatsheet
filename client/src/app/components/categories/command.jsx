import React from 'react';
import PropTypes from 'prop-types';

const Command = ({ command }) => (
  <React.Fragment>
    <div className="command-content">
      <div className="command-description">
        { command.description }
      </div>
      <code className="command-script">
        { command.script }
      </code>
    </div>
  </React.Fragment>
);

Command.propTypes = {
  command: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default Command;
