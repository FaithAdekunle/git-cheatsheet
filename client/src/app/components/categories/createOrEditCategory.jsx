import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Modal from '../common/modal';

class CreateOrEditCategory extends React.Component {
  constructor(props) {
    super(props);
    this.state = { category: { ...props.category }, errors: { commands: [] } };
    this.createOrEditCategory = this.createOrEditCategory.bind(this);
    this.onChange = this.onChange.bind(this);
    this.addCommand = this.addCommand.bind(this);
    this.removeCommand = this.removeCommand.bind(this);
    this.onCommandChange = this.onCommandChange.bind(this);
    this.validateCommand = this.validateCommand.bind(this);
    this.validate = this.validate.bind(this);
    this.abort = this.abort.bind(this);
  }

  onCommandChange(event, key) {
    const { category, errors } = this.state;
    const { target } = event;
    const changedCommand = category.commands.find(command => command.key === key);
    changedCommand[target.name] = target.value;
    if (errors.commands[changedCommand.key]) {
      errors.commands[changedCommand.key] = this.validateCommand(changedCommand);
    }
    this.setState({ category, errors });
  }

  onChange(event) {
    const { category, errors } = this.state;
    const { target } = event;
    if (target.name === 'privacyStatus') {
      category[target.name] = !target.checked;
    } else {
      category[target.name] = target.value;
      if (target.value) {
        errors[target.name] = '';
      } else {
        errors[target.name] = `${target.name} field cannot be empty`;
      }
    }
    this.setState({ category, errors });
  }

  abort() {
    const { category } = this.state;
    const { abort } = this.props;
    abort(category);
  }

  addCommand() {
    const { category } = this.state;
    category.commands.push({
      script: '',
      description: '',
      keywords: '',
      key: Date.now(),
    });
    this.setState({ category });
  }

  createOrEditCategory(event) {
    event.preventDefault();
    const valid = this.validate();
    if (valid) {
      const { createOrEditCategory } = this.props;
      const { category } = this.state;
      let validCategory = { ...category };
      if (validCategory._id) {
        validCategory = {
          _id: validCategory._id,
          title: validCategory.title,
        };
      } else {
        validCategory.commands = validCategory.commands.map(command => ({
          script: command.script,
          description: command.description,
          keywords: command.keywords.trim() ? command.keywords.trim().split(',') : [],
        }));
      }
      createOrEditCategory(validCategory);
    }
  }

  removeCommand(key) {
    const { category } = this.state;
    category.commands = category.commands.filter(command => command.key !== key);
    this.setState({ category });
  }

  validateCommand(command) {
    let error = '';
    if (!command.script.trim() && !command.description.trim()) {
      error = 'script and description fields cannot be empty';
    } else {
      ['script', 'description'].forEach((field) => {
        if (!command[field].trim()) error = `${field} field cannot be empty`;
      });
    }
    return error;
  }

  validate() {
    const { category, errors } = this.state;
    let valid = true;
    if (!category.title.trim()) {
      errors.title = 'title field cannot be empty';
      valid = false;
    }
    if (!category._id) {
      category.commands.forEach((command) => {
        const error = this.validateCommand(command);
        errors.commands[command.key] = error;
        valid = valid ? !error : valid;
      });
    }
    this.setState({ category, errors });
    return valid;
  }

  render() {
    const { category, errors } = this.state;
    return (
      <React.Fragment>
        <Modal>
          {
            <div className="add-or-edit-category">
              <form className="add-or-edit-category-form" onSubmit={this.createOrEditCategory}>
                <div className="form-group">
                  <label>Category Title *</label>
                  <span className="error">{errors.title || ''}</span>
                  <input
                    type="text"
                    value={category.title}
                    className="form-control title-field"
                    name="title"
                    onChange={this.onChange}
                  />
                </div>
                {
                  category._id ? '' : (
                    <React.Fragment>
                      <label>Git Command(s) *</label>
                      {
                        category.commands.map(command => (
                          <div className="command-input-group" key={command.key}>
                            <div className="category-command-error-action-wrapper">
                              <span className="error">
                                {errors.commands[command.key] || ''}
                              </span>
                              {
                                category.commands.length > 1 ? (
                                  <span
                                    className="remove-category-command-icon"
                                    onClick={() => this.removeCommand(command.key)}
                                  >
                                    <FontAwesomeIcon icon="minus-circle" />
                                  </span>
                                ) : ''
                              }
                            </div>
                            <div className="input-group">
                              <input
                                type="text"
                                className="form-control"
                                placeholder="git script here *"
                                value={command.script}
                                name="script"
                                onChange={event => this.onCommandChange(event, command.key)}
                              />
                              <input
                                type="text"
                                className="form-control"
                                placeholder="keywords separated by a comma"
                                value={command.keywords}
                                name="keywords"
                                onChange={event => this.onCommandChange(event, command.key)}
                              />
                              <input
                                type="text"
                                className="form-control"
                                placeholder="short description here *"
                                value={command.description}
                                name="description"
                                onChange={event => this.onCommandChange(event, command.key)}
                              />
                            </div>
                          </div>
                        ))
                      }
                      <div className="add-category-command-wrapper">
                        <span className="add-category-command-icon" onClick={this.addCommand}>
                          <FontAwesomeIcon icon="plus-circle" />
                        </span>
                      </div>
                    </React.Fragment>
                  )
                }
                {
                  category._id ? '' : (
                    <div className="form-check">
                      <input
                        className="form-check-input"
                        type="checkbox"
                        name="privacyStatus"
                        defaultChecked={!category.privacyStatus}
                        onChange={this.onChange}
                      />
                      <label className="form-check-label">
                        Public Status
                      </label>
                    </div>
                  )
                }
                <div className="underline" />
                <div className="text-right">
                  <button type="button" className="abort-add-or-edit-category-button" onClick={this.abort}>
                    Cancel
                  </button>
                  <button type="submit">
                    {`${category._id ? 'Edit' : 'Add'} Category`}
                  </button>
                </div>
              </form>
            </div>
          }
        </Modal>
      </React.Fragment>
    );
  }
}

CreateOrEditCategory.propTypes = {
  category: PropTypes.objectOf(PropTypes.any).isRequired,
  abort: PropTypes.func.isRequired,
  createOrEditCategory: PropTypes.func.isRequired,
};

export default CreateOrEditCategory;
