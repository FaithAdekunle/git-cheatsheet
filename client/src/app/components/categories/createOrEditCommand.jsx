import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Modal from '../common/modal';

class CreateOrEditCommand extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      commands: props.commands,
      errors: {},
    };
    this.onChange = this.onChange.bind(this);
    this.abort = this.abort.bind(this);
    this.validate = this.validate.bind(this);
    this.createOrEditCommand = this.createOrEditCommand.bind(this);
    this.validateCommand = this.validateCommand.bind(this);
    this.addCommand = this.addCommand.bind(this);
    this.removeCommand = this.removeCommand.bind(this);
  }

  onChange(event, key) {
    const { commands, errors } = this.state;
    const command = commands.find(command => command.key === key);
    const { target } = event;
    if (target.name === 'keywords') {
      command.keywords = target.value.split(',');
    } else {
      command[target.name] = target.value;
      if (errors[command.key]) this.validateCommand(command);
    }
    this.setState({ commands });
  }

  addCommand() {
    const { commands } = this.state;
    commands.push({
      script: '',
      description: '',
      keywords: [],
      key: Date.now(),
    });
    this.setState({ commands });
  }

  removeCommand(key) {
    const { commands } = this.state;
    this.setState({ commands: commands.filter(command => command.key !== key) });
  }

  validateCommand(command) {
    const { errors } = this.state;
    let error = '';
    let valid = true;
    if (!command.script.trim() && !command.description.trim()) {
      error = 'script and description fields cannot be empty';
      valid = false;
    } else {
      ['script', 'description'].forEach((field) => {
        if (!command[field].trim()) {
          error = `${field} field cannot be empty`;
          valid = false;
        }
      });
    }
    errors[command.key] = error;
    this.setState({ errors });
    return valid;
  }

  validate() {
    const { commands } = this.state;
    let valid = true;
    commands.forEach((command) => {
      const commandIsValid = this.validateCommand(command);
      if (valid && !commandIsValid) valid = false;
    });
    return valid;
  }

  createOrEditCommand(event) {
    event.preventDefault();
    const { createOrEditCommand, categoryId } = this.props;
    const { commands } = this.state;
    if (this.validate()) createOrEditCommand({ commands, categoryId });
  }

  abort() {
    const { abort, categoryId } = this.props;
    const { commands } = this.state;
    abort({ commands, categoryId });
  }

  render() {
    const { commands, errors } = this.state;
    const { categoryId } = this.props;
    return (
      <Modal>
        <div className="create-or-edit-command">
          <form className="create-or-edit-command-form" onSubmit={this.createOrEditCommand}>
            {
              commands.map(command => (
                <div className="command-input-group" key={command.key}>
                  <div className="command-error-action-wrapper">
                    <span className="error">
                      {errors[command.key] || ''}
                    </span>
                    {
                      commands.length > 1 ? (
                        <span
                          className="remove-command-icon"
                          onClick={() => this.removeCommand(command.key)}
                        >
                          <FontAwesomeIcon icon="minus-circle" />
                        </span>
                      ) : ''
                    }
                  </div>
                  <div className="input-group">
                    <input
                      type="text"
                      placeholder="git script here *"
                      value={command.script}
                      className="form-control"
                      name="script"
                      onChange={event => this.onChange(event, command.key)}
                    />
                    <input
                      type="text"
                      placeholder="keywords separated by a comma"
                      value={command.keywords.join(',')}
                      className="form-control"
                      name="keywords"
                      onChange={event => this.onChange(event, command.key)}
                    />
                    <input
                      type="text"
                      placeholder="short description here *"
                      value={command.description}
                      className="form-control"
                      name="description"
                      onChange={event => this.onChange(event, command.key)}
                    />
                  </div>
                </div>
              ))
            }
            {
              categoryId ? (
                <div className="add-command-wrapper">
                  <span className="add-command-icon" onClick={this.addCommand}>
                    <FontAwesomeIcon icon="plus-circle" />
                  </span>
                </div>
              ) : ''
            }
            <div className="underline" />
            <div className="text-right">
              <button
                type="button"
                className="abort-create-or-edit-command-button"
                onClick={this.abort}
              >
                Cancel
              </button>
              <button type="submit">
                {`${categoryId ? 'Add' : 'Edit'} Command`}
              </button>
            </div>
          </form>
        </div>
      </Modal>
    );
  }
}

CreateOrEditCommand.propTypes = {
  commands: PropTypes.arrayOf(PropTypes.any).isRequired,
  abort: PropTypes.func.isRequired,
  createOrEditCommand: PropTypes.func.isRequired,
  categoryId: PropTypes.any.isRequired,
};

export default CreateOrEditCommand;
