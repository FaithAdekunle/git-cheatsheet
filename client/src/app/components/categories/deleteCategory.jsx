import React from 'react';
import PropTypes from 'prop-types';
import Modal from '../common/modal';

const DeleteCategory = ({ category, abort, deleteCategory }) => (
  <React.Fragment>
    <Modal>
      <div className="delete-category">
        <h4>
          {`Category '${category.title.toUpperCase()}' has ${category.commands.length} command(s)`}
        </h4>
        <span>All commands in this category will also be removed...</span>
        <div className="underline" />
        <div className="text-right">
          <button type="button" className="abort-delete-category-button" onClick={() => abort()}>
            Abort
          </button>
          <button type="button" className="delete-category-button" onClick={() => deleteCategory()}>
            Delete anyway
          </button>
        </div>
      </div>
    </Modal>
  </React.Fragment>
);

DeleteCategory.propTypes = {
  category: PropTypes.objectOf(PropTypes.any).isRequired,
  abort: PropTypes.func.isRequired,
  deleteCategory: PropTypes.func.isRequired,
};

export default DeleteCategory;
