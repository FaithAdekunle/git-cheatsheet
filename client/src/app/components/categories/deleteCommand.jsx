import React from 'react';
import PropTypes from 'prop-types';
import Modal from '../common/modal';

const DeleteCommand = ({ command, abort, deleteCommand }) => (
  <React.Fragment>
    <Modal>
      <div className="delete-command">
        <h4>
          Delete
          {'    "'}
          <code>{command.script}</code>
          {'" command?'}
        </h4>
        <div className="underline" />
        <div className="text-right">
          <button type="button" className="abort-delete-command-button" onClick={abort}>
            Abort
          </button>
          <button type="button" className="delete-command-button" onClick={deleteCommand}>
            Delete
          </button>
        </div>
      </div>
    </Modal>
  </React.Fragment>
);

DeleteCommand.propTypes = {
  command: PropTypes.objectOf(PropTypes.any).isRequired,
  abort: PropTypes.func.isRequired,
  deleteCommand: PropTypes.func.isRequired,
};

export default DeleteCommand;
