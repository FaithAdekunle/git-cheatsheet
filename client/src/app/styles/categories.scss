.categories-component {
  padding: 1em;
  button {
    cursor: pointer;
  }
  .categories-actions {
    display: flex;
    justify-content: space-between;
  }
  .category-components {
    .category-component {
      margin-top: 1.5em;
      .category-header {
        background-color: #31363F;
        padding: 0.2em 0.5em;
        font-size: 1em;
        cursor: pointer;
      }
      .category-body {
        max-height: 0;
        overflow: hidden;
        transition-property: max-height;
        transition-duration: 0.7s;
        transition-delay: 0s;
        color: #252a2e;
        .category-actions {
          display: flex;
          justify-content: space-between;
          padding: 0 0.3em;
          background-color: white;
          text-align: right;
          .fa-pen, .fa-trash, .privacy, .fa-plus {
            margin: auto 0 auto 0.7em;
            cursor: pointer;
            font-size: 0.7em;
            display: inline-block;
          }
        }
        .command-list {
          list-style-type: none;
          padding: 0;
          .command-list-item {
            background-color: #ffffe9;
            margin-top: 0.5em;
            margin-bottom: 0.5em;
            .command-actions {
              padding: 0 0.3em;
              text-align: right;
              font-size: 0.7em;
              background-color: #e5e5d1;
              span {
                cursor: pointer;
                &:hover {
                  text-decoration: underline;
                }
              }
            }
            .command-content {
              padding: 0 0.5em 0.5em 0.5em;
              .command-description {
                font-size: 0.85em;
              }
              .command-script {
                font-size: 0.7em;
                cursor: pointer;
              }
            }
          }
        }
      }
      .expand {
        max-height: 50em;
        overflow-y: scroll;
      }
    }
  }
}
.delete-category, .add-or-edit-category, .create-or-edit-command {
  padding: 1em;
  background-color: #252a2e;
  .underline {
    height: 0.1em;
    background-color: #31363F;
    margin: 1.2em 0 0.5em 0;
  }
  button {
    display: inline-block;
    margin-left: 1em;
  }
  label {
    margin-bottom: 0.3rem;
  }
  input {
     border-radius: unset;
   }
  input[type="submit"] {
    cursor: pointer;
  }
  .error {
    display: block;
    color: #f66;
    font-size: 0.7rem;
  }
}
.add-or-edit-category {
  .form-check-label {
    margin-top: 0.13rem;
  }
  .command-input-group {
    padding-bottom: 0.5em;
    border-top: 1px solid #7F7F7F;
    border-bottom: 1px solid #7F7F7F;
    margin-bottom: 0.5em;
    input {
      width: 50%;
    }
    .category-command-error-action-wrapper {
      display: flex;
      justify-content: space-between;
      .remove-category-command-icon {
        cursor: pointer;
        font-size: 0.8em;
      }
    }
  }
  .add-category-command-wrapper {
    text-align: center;
    .add-category-command-icon {
      cursor: pointer;
      font-size: 1em;
    }
  }
}
.create-or-edit-command {
  .command-input-group {
    padding-bottom: 0.5em;
    border-top: 1px solid #7F7F7F;
    border-bottom: 1px solid #7F7F7F;
    margin-bottom: 0.5em;
    input {
      width: 50%;
    }
    .command-error-action-wrapper {
      display: flex;
      justify-content: space-between;
      .remove-command-icon {
        cursor: pointer;
        font-size: 0.8em;
      }
    }
  }
  .add-command-wrapper {
    text-align: center;
    .add-command-icon {
      cursor: pointer;
      font-size: 1em;
    }
  }
}
