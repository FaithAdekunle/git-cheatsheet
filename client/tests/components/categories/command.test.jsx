import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Command from '../../../src/app/components/categories/command';
import categories from '../../categories';

const commandProps = {
  command: categories[0].commands[0],
};

describe('<Command />', () => {
  test('should match snapshot', () => {
    const wrapper = shallow(<Command {...commandProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
