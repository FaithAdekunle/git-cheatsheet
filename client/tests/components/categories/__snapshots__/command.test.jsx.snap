// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Command /> should match snapshot 1`] = `
<Fragment>
  <div
    className="command-content"
  >
    <div
      className="command-description"
    >
      Install git on macOS with Homebrew
    </div>
    <code
      className="command-script"
    >
      brew install git
    </code>
  </div>
</Fragment>
`;
